src/cascadia/TerminalControl/ControlCore.cpp

```diff
@@ -1951,6 +1951,40 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         return hstring{ str };
     }
 
+    // Get all of our recent commands. This will only really work if the user has enabled shell integration.
+    Control::CommandHistoryContext ControlCore::CommandHistory() const
+    {
+        auto terminalLock = _terminal->LockForWriting();
+        auto context = winrt::make_self<CommandHistoryContext>();
+        const auto& textBuffer = _terminal->GetTextBuffer();
+
+        for (const auto& mark : _terminal->GetScrollMarks())
+        {
+            // The command text is between the `end` (which denotes the end of
+            // the prompt) and the `commandEnd`.
+            bool markHasCommand = mark.commandEnd.has_value() &&
+                                  mark.commandEnd != mark.end;
+            if (!markHasCommand)
+            {
+                continue;
+            }
+
+            // Get the text of the command
+            const auto line = mark.end.y;
+            const auto& row = textBuffer.GetRowByOffset(line);
+            const auto commandText = row.GetText(mark.end.x, mark.commandEnd->x);
+
+            // Trim off trailing spaces.
+            const auto strEnd = commandText.find_last_not_of(UNICODE_SPACE);
+            if (strEnd != std::string::npos)
+            {
+                const auto trimmed = commandText.substr(0, strEnd + 1);
+                context->History().Append(winrt::hstring{ trimmed });

```

Comment:
The call to `History()` can be hoisted out of the loop.
But what would be even better IMHO was a `ReplaceHistory` method. That way you could build a fresh `std::vector` here and replace the entire vector over there at once. You could even make it a constructor argument and since we got access to the implementation we can even pass the `std::vector` as is.
I'm not sure how I should express why I think it would be better, but you remember the number of bugs we had in the settings UI with the navigation item vector? 😅 I think we should try to avoid such design.